File: oldsrc/src/main/java/old/activities/main/fragments/shows/ShowsFragment.java
package jdr.tvtracker.activities.main.fragments.shows;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import jdr.tvtracker.R;
import jdr.tvtracker.activities.main.MainActivityViewModel;
import jdr.tvtracker.data.entities.ShowBasic;
import jdr.tvtracker.utils.SpacingDecoration;

public class ShowsFragment extends Fragment {
    private MainActivityViewModel mainActivityViewModel;
    private ShowsFragmentAdapter showsFragmentAdapter;

    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.default_fragment, container, false);
    }

    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (getActivity() != null) {
            this.mainActivityViewModel = (MainActivityViewModel) ViewModelProviders.of(getActivity()).get(MainActivityViewModel.class);
        }
        RecyclerView recyclerView = (RecyclerView) view.findViewById(R.id.recyclerView);
        this.showsFragmentAdapter = new ShowsFragmentAdapter(view.getContext());
        recyclerView.setLayoutManager(new GridLayoutManager(view.getContext(), calculateNumberOfColumns(view.getContext())));
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        int spacing = getResources().getDimensionPixelSize(R.dimen.margin);
        recyclerView.addItemDecoration(new SpacingDecoration(spacing, spacing));
        recyclerView.setAdapter(this.showsFragmentAdapter);
        observeLiveData();
    }

    private int calculateNumberOfColumns(Context context) {
        DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
        int columnCount = (int) ((((float) displayMetrics.widthPixels) / displayMetrics.density) / ((float) 160));
        return columnCount > 2 ? columnCount : 2;
    }

    private void observeLiveData() {
        this.mainActivityViewModel.getShowList().observe(this, new Observer<List<ShowBasic>>() {
            public void onChanged(@Nullable List<ShowBasic> showList) {
                if (showList != null && showList.size() > 0) {
                    ShowsFragment.this.showsFragmentAdapter.updateShowsList(showList);
                }
            }
        });
    }
}

File: feature/shows/src/main/java/jdr/tv/feature/shows/ui/ShowsFragment.kt
package jdr.tv.feature.shows.ui

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import com.squareup.cycler.Recycler
import com.squareup.cycler.toDataSource
import jdr.tv.common.navigation.GlobalActions
import jdr.tv.common.ui.extensions.bind
import jdr.tv.common.ui.extensions.displayMetrics
import jdr.tv.common.ui.extensions.dpToPixels
import jdr.tv.common.ui.extensions.loadPoster
import jdr.tv.common.ui.extensions.setupToolbar
import jdr.tv.common.ui.extensions.span
import jdr.tv.common.ui.utils.SpacingItemDecoration
import jdr.tv.data.core.di.DataComponent
import jdr.tv.data.local.entities.Show
import jdr.tv.feature.shows.databinding.FragmentShowsBinding
import jdr.tv.feature.shows.databinding.ItemShowPosterBinding
import jdr.tv.feature.shows.di.inject
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import javax.inject.Inject
import kotlin.math.roundToInt

class ShowsFragment @Inject constructor(private val component: DataComponent) : Fragment() {

    companion object {
        private const val SPACING = 16
        private const val SPAN = 200
    }

    private var binding: FragmentShowsBinding? = null
    private var recycler: Recycler<Show>? = null

    @Inject
    lateinit var viewModel: ShowsViewModel

    override fun onAttach(context: Context) {
        super.onAttach(context)
        inject(component)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(jdr.tv.common.navigation.R.menu.menu_main, menu)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return FragmentShowsBinding.inflate(inflater, container, false).run {
            binding = this
            recyclerView.span = calculateSpan()
            root
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        setHasOptionsMenu(true)
        setupToolbar(jdr.tv.common.ui.R.id.toolbar, jdr.tv.common.navigation.R.string.shows)
        setupRecyclerView()
        observe()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
        recycler = null
    }

    private fun calculateSpan(): Int {
        return (requireContext().displayMetrics().widthPixels.toFloat() / requireContext().dpToPixels(SPAN)).roundToInt()
    }

    private fun setupRecyclerView() = binding?.recyclerView?.apply {
        recycler = Recycler.adopt(this) {
            stableId(Show::id)
            row<Show, View> {
                bind(ItemShowPosterBinding::inflate) { show ->
                    posterImageView.loadPoster(show.posterPath)
                    posterImageView.setOnClickListener {
                        navigate(show.id)
                    }
                }
            }
        }
        addItemDecoration(SpacingItemDecoration.GridLayout(requireContext().dpToPixels(SPACING)))
    }

    private fun observe() {
        lifecycleScope.launch {
            viewModel.addedShowList.collect {
                recycler?.update { data = it.toDataSource() }
            }
        }
    }

    private fun navigate(showId: Long) {
        findNavController().navigate(GlobalActions.actionDetails(showId))
    }
}
